src/main/java/de/zalando/aruha/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -0,0 +1,263 @@
+            log.info("Incrementing topology version");
+            final Integer curVersion = Integer.parseInt(new String(curatorFramework.getData().forPath(getSubscriptionPath("/topology")), CHARSET));
+            curatorFramework.setData().forPath(getSubscriptionPath("/topology"), String.valueOf(curVersion + 1).getBytes(CHARSET));
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public Partition[] listPartitions() {
+        log.info("fetching partitions information");
+        try {
+            final List<Partition> partitions = new ArrayList<>();
+            for (final String topic : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics"))) {
+                for (final String partition : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics/" + topic))) {
+                    final Partition.PartitionKey key = new Partition.PartitionKey(topic, partition);
+                    partitions.add(deserializeNode(key, curatorFramework.getData().forPath(getPartitionPath(key))));
+                }
+            }
+            return partitions.toArray(new Partition[partitions.size()]);
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public Session[] listSessions() {
+        log.info("fetching sessions information");
+        final List<Session> sessions = new ArrayList<>();
+        try {
+            for (final String sessionId : curatorFramework.getChildren().forPath(getSubscriptionPath("/sessions"))) {
+                final int weight = Integer.parseInt(new String(curatorFramework.getData().forPath(getSubscriptionPath("/sessions/" + sessionId)), CHARSET));
+                sessions.add(new Session(sessionId, weight));
+            }
+            return sessions.toArray(new Session[sessions.size()]);
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public ZKSubscription subscribeForSessionListChanges(final Runnable listener) {
+        log.info("subscribeForSessionListChanges: " + listener.hashCode());
+        return ChangeListener.forChildren(curatorFramework, getSubscriptionPath("/sessions"), listener);
+    }
+
+    @Override
+    public ZKSubscription subscribeForTopologyChanges(final Runnable onTopologyChanged) {
+        log.info("subscribeForTopologyChanges");
+        return ChangeListener.forData(curatorFramework, getSubscriptionPath("/topology"), onTopologyChanged);
+    }
+
+    @Override
+    public ZKSubscription subscribeForOffsetChanges(final Partition.PartitionKey key, final Runnable commitListener) {
+        log.info("subscribeForOffsetChanges");
+        return ChangeListener.forData(curatorFramework, getPartitionPath(key) + "/offset", commitListener);
+    }
+
+    @Override
+    public long getOffset(final Partition.PartitionKey key) {
```

Comment:
Because all the code that is written in StreamingState and PartitionData uses this functionality now (it uses offsets to calculate sent data size).
I decided not to make current solution overcomplicated.